integration_tests/websocket_test.go

```diff
@@ -0,0 +1,222 @@
+
+	//websocket server
+	h := WsHandler{handler: controllers.NewInvoiceStreamController(suite.service).StreamInvoices}
+	server := httptest.NewServer(http.HandlerFunc(h.ServeHTTP))
+	suite.websocketServer = server
+	suite.wsUrl = "ws" + strings.TrimPrefix(suite.websocketServer.URL, "http") + fmt.Sprintf("?token=%s", suite.userToken)
+	suite.wsUrl2 = "ws" + strings.TrimPrefix(suite.websocketServer.URL, "http") + fmt.Sprintf("?token=%s", suite.userToken2)
+}
+
+func (suite *WebSocketTestSuite) TestWebSocket() {
+
+	//start listening to websocket
+	ws, _, err := websocket.DefaultDialer.Dial(suite.wsUrl, nil)
+	assert.NoError(suite.T(), err, err)
+	_, msg, err := ws.ReadMessage()
+	assert.NoError(suite.T(), err, err)
+	keepAlive := KeepAlive{}
+	err = json.Unmarshal([]byte(msg), &keepAlive)
+	assert.NoError(suite.T(), err, err)
+	assert.Equal(suite.T(), "keepalive", keepAlive.Type)
+
+	// create incoming invoice and fund account
+	invoice := suite.createAddInvoiceReq(1000, "integration test websocket 1", suite.userToken)
+	sendPaymentRequest := lnrpc.SendRequest{
+		PaymentRequest: invoice.PayReq,
+		FeeLimit:       nil,
+	}
+	_, err = suite.fundingClient.SendPaymentSync(context.Background(), &sendPaymentRequest)
+	assert.NoError(suite.T(), err)
+
+	_, msg, err = ws.ReadMessage()
+	assert.NoError(suite.T(), err, err)
+	event := ExpectedInvoiceEventWrapper{}
+	err = json.Unmarshal([]byte(msg), &event)
+	assert.NoError(suite.T(), err, err)
+	assert.Equal(suite.T(), event.Type, "invoice")
+	assert.Equal(suite.T(), int64(1000), event.Invoice.Amount)
+	assert.Equal(suite.T(), "integration test websocket 1", event.Invoice.Description)
+	//create 2nd subscription, create invoice, pay invoice, assert that invoice is received twice
+	//start listening to websocket
+	ws2, _, err := websocket.DefaultDialer.Dial(suite.wsUrl, nil)
+	assert.NoError(suite.T(), err, err)
+	//read keepalive msg
+	_, _, err = ws2.ReadMessage()
+	assert.NoError(suite.T(), err, err)
+	invoice = suite.createAddInvoiceReq(1000, "integration test websocket 2", suite.userToken)
+	sendPaymentRequest = lnrpc.SendRequest{
+		PaymentRequest: invoice.PayReq,
+		FeeLimit:       nil,
+	}
+	_, err = suite.fundingClient.SendPaymentSync(context.Background(), &sendPaymentRequest)
+	assert.NoError(suite.T(), err)
+	_, msg1, err := ws.ReadMessage()
+	assert.NoError(suite.T(), err, err)
+	_, msg2, err := ws2.ReadMessage()
+	assert.NoError(suite.T(), err)
+
+	event1 := ExpectedInvoiceEventWrapper{}
+	err = json.Unmarshal([]byte(msg1), &event1)
+	assert.NoError(suite.T(), err)
+	event2 := ExpectedInvoiceEventWrapper{}
+	err = json.Unmarshal([]byte(msg2), &event2)
+	assert.NoError(suite.T(), err, err)
+	assert.Equal(suite.T(), "integration test websocket 2", event1.Invoice.Description)
+	assert.Equal(suite.T(), "integration test websocket 2", event2.Invoice.Description)
+	//close 1 subscription, assert that the existing sub still receives their invoices
+	ws.Close()
```

Comment:
is there something else to be tested when this is closed, maybe pubsub? if not this is ok for me

Reply:
I don't know how to test that it was actually removed from the subscribers, the PubSub as it's written now does not expose the number of subscribers.